### PR TITLE
[WIP] For changing the initialization mechanism 

### LINKED ISSUE #970: Metaissue: Improve initialization
@memimo , this is going to be you CCW task. I don't know what's need to be done yet, I just summarize information.

I list here the issues with some summary because the title doesn't always correspond to the discussion.  
I found several discussions on initialization:
- #773 where many people are complaining about the current way of initialization and pushing the config
- #740 which proposes a new way to initialize stuff
- #384 another thread on initialization
- #31 and yet another

Other threads which concern initialization:
- #347 discussion on where we should check if every brick was initialized
- #463 enhance initialization schemes printing
- #816 argues about noninitializable sequence and raises a question how we should manage several different properties
- #935 figures out that allocation and possibly initialization is broken with several parent bricks

@rizar , I forgot something?

### PR DESCRIPTION
**Pull request by memimo**
fixes #970
This is just proof of concept. When you guys agree that this is the way to go, I will change all the codes affected by this, plus rebase to the head of the master.
@rizar @dmitriy-serdyuk

### CODE REVIEW

**Comment by dwf** [comment]
I thought that part of the the plan was basically to be able to examine
kwargs for anything ending in `_init`, which chops off the `_init` and
looks that string up as a role name?

File: blocks/bricks/interfaces.py [hunk 1]
```diff
@@ -141,31 +141,61 @@ class Initializable(RNGMixin, Brick):
 
     """
     has_biases = True
+    initialization_roles = set([])
 
     @lazy()
-    def __init__(self, weights_init=None, biases_init=None, use_bias=True,
-                 seed=None, **kwargs):
+    def __init__(self,
```
Comment (rizar) [inline blocks/bricks/interfaces.py]: Please, do not use one-keyword-argument-per-line indentation scheme. It is never used in Blocks, for the sake of maintaining a common style for the codebase, I would like to ask you not to use it as well.

File: blocks/bricks/interfaces.py [hunk 2]
```diff
@@ -141,31 +141,61 @@ class Initializable(RNGMixin, Brick):
 
     """
     has_biases = True
+    initialization_roles = set([])
```
Comment (rizar) [inline blocks/bricks/interfaces.py]: What do you think about calling this one `parameter_roles`?

File: blocks/bricks/interfaces.py [hunk 3]
```diff
@@ -141,31 +141,61 @@ class Initializable(RNGMixin, Brick):
...
             for child in self.children:
                 if (isinstance(child, Initializable) and
                         hasattr(child, 'biases_init')):
                     child.biases_init = self.biases_init
+        # import ipdb; ipdb.set_trace()
 
+    def _collect_roles(self):
+        for child in self.children:
+            if isinstance(child, Initializable):
+                self.initialization_roles.update(child.initialization_roles)
```
Comment (rizar) [inline blocks/bricks/interfaces.py]: We've got a big problem here: you defined `initialization_roles` as a class attribute. Now this assignment will affect all the object of the class, which I guess is not what you wanted.

File: blocks/bricks/interfaces.py [hunk 4]
```diff
@@ -141,31 +141,61 @@ class Initializable(RNGMixin, Brick):
 
     """
     has_biases = True
```
Comment (rizar) [inline blocks/bricks/interfaces.py]: Why do we need `has_biases` when we have `initialization_roles`?

**Comment by rizar** [comment]
@dwf , @bartvm , I guess I need a Python advice here. Is there a general agreement on when one should call `super().__init__`, in the beginning or in the end of `def __init__`? 

I think that in this PR we should do the following: first initialize the list of `children`, then call `super().__` that will eventually call `Initializable.__init__`. In the latter the roles of the children will be collected in order to validate the initialization schemes provided by the user.

This would only work if everybody places their `super().__init__` call in the end of `def __init__`. And I have a bad feeling about the numerous facepalms that this can cause. What do you think?

**Comment by rizar** [comment]
@memimo , nice to see that you started working on the issue, but this PR is definitely very raw. We have to work closer together to come up with a good and implementable design.

File: blocks/bricks/interfaces.py (hunk 4 above)
Comment (memimo) [inline blocks/bricks/interfaces.py]: backward compatibility. I know there was a discussion about dealing with biases like rest of parameters but it didn't look as it was conclusive. I can get rid of it, and treat backward compatibility with a replication warning.

File: blocks/bricks/interfaces.py (hunk 2 above)
Comment (memimo) [inline blocks/bricks/interfaces.py]: yeah, that could work.

File: blocks/bricks/interfaces.py (hunk 1 above)
Comment (memimo) [inline blocks/bricks/interfaces.py]: yeah, will go thorough formatting issues at the end. This was just to open the discussion about the way to solve this issues.

File: blocks/bricks/interfaces.py (hunk 4 above)
Comment (rizar) [inline blocks/bricks/interfaces.py]: Backward compatibility is going to kill us, if we go for it. We have stable versions for those who work with their old models. So I think we should care only about getting a new solution for initialization.

**Comment by rizar** [comment]
@dwf , @bartvm , can you please comment on what I wrote above?

File: blocks/bricks/interfaces.py (hunk 4 above)
Comment (dwf) [inline blocks/bricks/interfaces.py]: I complained about the name `has_biases` in #697 and would be happy to see it go.

**Comment by dwf** [comment]
Whether you put `super().__init__(...)` in the beginning, end, or middle of the `__init__` definition (applies to any super call in any method, really) depends on what your constructor does and what you are relying on the superclass to do.

Sometimes you need to compute arguments for the `super().__init__` call based on your own arguments, so it's naturally going to come after that. Sometimes you want to let `super().__init__` configure something that you then tinker with slightly, so it's naturally going to come before that.

If there's no interaction between what I'm doing and what the superclass is doing then I typically put the call at the end, but I think this is a matter of taste. There are some [subtleties with respect to super and multiple inheritance](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/) but the way we use it, they usually don't pose a problem as long as everyone calls super.

**Comment by rizar** [comment]
OK, thanks for the feedback. 

I think I understood where the confusion comes from: we do not allow children to be passed as an argument to `Brick`. That's why it always initializes the list of chiden with `[]`, and that's why it is becomes crucial to fill `self.children` with content only after the super call. This should be changed, and the recommended way of setting the children should be by passing them as an argument to `Brick`. If we go this way, `Initializable.__init__` will call `super().__init__` first, which will initialize the list of children, which in turn allows it to check which initialization schemes a brick can have.

File: blocks/bricks/interfaces.py (hunk 4 above)
Comment (rizar) [inline blocks/bricks/interfaces.py]: Agreed.

**Comment by memimo** [comment]
should I make a separate PR to fix the children issue first before continuing with this PR?

**Comment by rizar** [comment]
If you agree that this is the way to go, then yes. I will create a ticket.

**Comment by memimo** [comment]
Yeah, I think this makes things more clear and avoid the __init__ order issue.

**Pull request closed without merging**